classes/SObjectRepository.cls

```diff
@@ -0,0 +1,286 @@
+        }
+
+        return this;
+    }
+
+    public List<String> getSelectedFields() {
+        return this.selectedFields;
+    }
+
+    public SObjectRepository selectChild(String relationshipName, ISObjectRepository childRepo) {
+        schemaValidator.validateChildRelationship(relationshipName, childRepo.getSObjectType());
+        String childQueryString = childRepo.getQueryString();
+        String originalFromClause = String.format(
+            FROM_OBJECT_TEMPLATE,
+            new List<String> {
+                childRepo
+                    .getSObjectType()
+                    .getDescribe()
+                    .getName()
+            }
+        );
+        String newFromClause = String.format(FROM_OBJECT_TEMPLATE, new List<String> { relationshipName });
+        String newChildQueryString = childQueryString.replace(originalFromClause, newFromClause);
+
+        if(String.isBlank(newChildQueryString.substringAfter('WHERE'))) {
+            newChildQueryString = newChildQueryString.substringBefore(' WHERE');
+        }
+
+        this.selectedChildQueries.add(newChildQueryString);
+
+        return this;
+    }    
+
+    public SObjectRepository selectParent(Schema.SObjectField relationshipField, ISObjectRepository parentRepo) {
+        schemaValidator.validateParentRelationship(relationshipField, parentRepo.getSObjectType());
+        String relationshipName = relationshipField.getDescribe().getRelationshipName();
+        for (String currParentField : parentRepo.getSelectedFields()) {
+            this.selectedFields.add(relationshipName + '.' + currParentField);
+        }
+
+        return this;
+    }    
+
+    public SObjectRepository whereField(String fieldName, QueryCondition.Operator operator, Object value) {
+        schemaValidator.validateField(fieldName);
+        SObjectField field = this.objectType.getDescribe().fields.getMap().get(fieldName);
+        this.whereConditions.add(new QueryCondition(field, operator, value));
+        return this;
+    }
+
+    public SObjectRepository whereField(SObjectField field, QueryCondition.Operator operator, Object value) {
+        this.whereConditions.add(new QueryCondition(field, operator, value));
+        return this;
+    }
+
+    public SObjectRepository whereField(String fieldName, QueryCondition.Operator operator, List<Object> values) {
+        schemaValidator.validateField(fieldName);
+        SObjectField field = this.objectType.getDescribe().fields.getMap().get(fieldName);
+        this.whereConditions.add(new QueryCondition(field, operator, values));
+        return this;
+    }
+
+    public SObjectRepository whereField(SObjectField field, QueryCondition.Operator operator, List<Object> values) {
+        this.whereConditions.add(new QueryCondition(field, operator, values));
+        return this;
+    }
+
+    public SObjectRepository whereParent(Schema.SObjectField relationshipField, ISObjectRepository parentRepo) {
```

Comment:
Is there an example of when I'd use this vs. when I'd use `.joinParent`?

Reply:
per one of my other comments, it's a bit redundant to have both this and joinParent. The original idea was to have separate methods for selected fields on parent record vs filtering on fields on the parent record. joinParent handles both, so there's not really a good reason to use `.whereParent` instead of `.joinParent`